core/embed/extmod/modtrezorui/buffers.c

```diff
@@ -0,0 +1,75 @@
+/*
+ * This file is part of the Trezor project, https://trezor.io/
+ *
+ * Copyright (c) SatoshiLabs
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "buffers.h"
+#include "common.h"
+#include "fonts/fonts.h"
+#include "memzero.h"
+
+#if USE_DMA2D
+
+#if defined BOOTLOADER
+#define BUFFER_SECTION __attribute__((section(".buf")))
+#else
+#define BUFFER_SECTION
+#endif
+
+#define BUFFERS_16BPP 3
+#define BUFFERS_4BPP 3
+#define BUFFERS_TEXT 1
+
+const int32_t text_buffer_height = FONT_MAX_HEIGHT;
+const int32_t buffer_width = DISPLAY_RESX;
+
+BUFFER_SECTION uint32_t line_buffer_16bpp[BUFFERS_16BPP][BUFFER_PIXELS / 2];
+BUFFER_SECTION uint32_t line_buffer_4bpp[BUFFERS_4BPP][BUFFER_PIXELS / 8];
+BUFFER_SECTION uint32_t
+    text_buffer[BUFFERS_TEXT][(FONT_MAX_HEIGHT * DISPLAY_RESX) / 8];
+
+uint8_t* buffers_get_line_buffer_16bpp(uint16_t idx, bool clear) {
```

Comment:
tried to do this in https://github.com/trezor/trezor-firmware/commit/afd4fc53a3ce7a3a87fd7c58dd203be1beba78da

not entirely sure about that though, seems a bit more convoluted solution. 

hope that `&mut *ptr` can't move the buffer?

(used this fixup to aslo separate buffers.rs from dma2d.rs, as it doesn't fit there much)

Reply:
And i gave up on using `FONT_MAX_HEIGHT` to determine text buffer size, instead using fixed 24 with an error if some font doesn't fit. The reason is that Rust bindgen doesn't see `FONT_MAX_HEIGHT`.

Reply:
see followup comments.

in general LGTM
It's more verbose on the C side, i'm kind of surprised that you have to cast to `uint8_t*`, but we could avoid that by `#define`ing the type instead of typedefing. I don't see a strong argument for either way, so either is fine by me.